### PR TITLE
Implementing #57: Real coefficients

### PR DESCRIPTION
**Pull request by Wikunia**
First go of implementing real coefficients. 
- [x] Check everything again
- [x] Check that there is no performance regression
- [x] Is it reasonable to support Rational weights in a different way?

### CODE REVIEW

**Comment by matbesancon** [comment]
To stay efficient when using non-concrete types, better parameterize with a generic:

```julia
mutable struct LinearVariables{T <: Real}
    indices             :: Vector{Int}
    coeffs              ::  Vector{T}
end 
```

Instead of Vector{Real}

**Comment by matbesancon** [comment]
When creating a LinearConstraint, either promote every real to the same type, or use different types for the different field types:
```julia
    mins                :: Vector{TM1}
    maxs                :: Vector{TM2}
    pre_mins            :: Vector{TP1}
    pre_maxs            :: Vector{TP1l} 
  # etc
```

File: test/small_eq_sum_real.jl [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    @variable(m, 0 <= max_val <= 10, Int)
+    @constraint(m, sum(weights.*x) == 0.3*max_val)
+    @objective(m, Max, max_val)
+    optimize!(m)
+    @test JuMP.termination_status(m) == MOI.OPTIMAL
+    @test JuMP.objective_value(m) == 2
+    @test JuMP.value.(x) == [0,1,1,0]
+end
+
+end
```
Comment (matbesancon) [inline test/small_eq_sum_real.jl]:
```suggestion
end

```

File: test/runtests.jl [hunk 2]
```diff
@@ -11,6 +11,7 @@ include("moi.jl")
 
 include("sudoku_fcts.jl")
 include("small_special_tests.jl")
+include("small_eq_sum_real.jl")
 include("sudoku_tests.jl")
 include("killer_sudoku_tests.jl")
 include("graph_color_tests.jl")
```
Comment (matbesancon) [inline test/runtests.jl]:
```suggestion
include("graph_color_tests.jl")

```

File: src/util.jl [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+function get_safe_lower_threshold(com::CS.CoM, val, divider)
+    ceil_threshold = cld(val, divider)
+    float_threshold = val/divider
+    threshold = convert(Int, ceil_threshold)
+    # if the difference is almost 1 we round in the other direction to provide a safe lower bound
+    if isapprox(ceil_threshold-float_threshold, 1.0; rtol=com.options.rtol, atol=com.options.atol)
+        threshold = convert(Int, fld(val, divider))
+    end
+    return threshold
+end
```
Comment (matbesancon) [inline src/util.jl]:
```suggestion
end

```

**Comment by Wikunia** [comment]
Thanks for having a look: 
I only need the linear variables if JuMP isn't used as the modelling layer, therefore I'm not sure whether it's reasonable to longer support that anyway. Maybe you have some thoughts on that. If JuMP is used all variables in a constraint will have the same type which is currently: Float64 anyway.

**Comment by matbesancon** [comment]
`LinearVariables` might be a tricky name, maybe `LinearCombination`, or `LinearFunction`? I think you should not hard-code `Float64`, JuMP will eventually get generic numerical types: https://github.com/JuliaOpt/JuMP.jl/issues/2025
Although it's not on the immediate roadmap I believe

File: src/options.jl [hunk 4]
```diff
@@ -3,15 +3,19 @@ mutable struct SolverOptions
     max_bt_steps        :: Int64
     backtrack_sorting   :: Bool
     keep_logs           :: Bool
+    rtol                :: Float64
+    atol                :: Float64
 end
 
 function get_default_options()
```
Comment (matbesancon) [inline src/options.jl]: This can be a default constructor `SolverOptions`

**Comment by matbesancon** [comment]
> I only need the linear variables if JuMP isn't used as the modelling layer

I think it is worth keeping, given the mention above that JuMP does not yet support generic numeric types

**Comment by Wikunia** [comment]
> When creating a LinearConstraint, either promote every real to the same type, or use different types for the different field types:
What is the easiest way to get the common type of values?
Thought Julia would do that automatically but looks like it doesn't.

**Comment by matbesancon** [comment]
> What is the easiest way to get the common type of values?

```julia
# on values
julia> promote(3, false)
(3, 0)

# on types
julia> promote_type(Float32, Float64)
Float64
```

File: src/ConstraintSolver.jl [hunk 5]
```diff
@@ -804,22 +827,27 @@ function backtrack!(com::CS.CoM, max_bt_steps; sorting=true)
...
-            end
-            push!(com.solutions, backtrack_obj.idx)
-            if com.best_sol == com.best_bound
-                return :Solved
-            else
-                # set all nodes to :Worse if they can't achieve a better solution
-                for bo in backtrack_vec
-                    if bo.status == :Open && obj_factor*bo.best_bound >= com.best_sol
-                        bo.status = :Worse
+                if com.best_sol == com.best_bound
```
Comment (matbesancon) [inline src/ConstraintSolver.jl]:
```julia
if com.best_sol == com.best_bound
    return :Solved
end
```

given that there is a return, no need for a else clause

**Comment by Wikunia** [comment]
Implemented:
- `LinearVariables` -> `LinearCombination`
- Generic type for `LinearCombination`

File: src/ConstraintSolver.jl [hunk 6]
```diff
@@ -83,26 +83,60 @@ struct NotEqualSet{T} <: MOI.AbstractScalarSet
...
+mutable struct LinearCombination{T <: Real}
     indices             :: Vector{Int}
-    coeffs              :: Vector{Int}
+    coeffs              :: Vector{T}
 end
 
-mutable struct LinearConstraint <: Constraint
+mutable struct LinearConstraint{T <: Real} <: Constraint
     idx                 :: Int
     fct                 :: Function
```
Comment (matbesancon) [inline src/ConstraintSolver.jl]: This may be for another PR but do you need to store the function here?

Comment (Wikunia) [reply src/ConstraintSolver.jl]: Currently every constraint has a function attached to it to call the corresponding constraint functions (pruning and feasibility check) later. 
- Is this unusual or not advised?
- What would be other reasonable ways to link the constraint type to the functions then?
Open for suggestions :)

Comment (matbesancon) [reply src/ConstraintSolver.jl]: Ok I see thanks for clarifying. So it's a bit unusual, but also expensive. The field of the struct is not concretely typed, just like the elements of a `Vector{Real}`. This yields to boxing and dynamic dispatch every time (slower). I think I will think of a proposal in a separate PR / issue

Comment (Wikunia) [reply src/ConstraintSolver.jl]: Hmm okay makes sense. Maybe having a subscriber model on `ConstraintType` and fields here `operator`? But yeah have to think about that longer.

Comment (matbesancon) [reply src/ConstraintSolver.jl]: basically have your constraint type implement a method `evaluate_value` and/or `is_valid`.
From there LinearConstraint would just be
`is_valid(lc::LinearConstraint, xvalues) = a^T xvalues <= b` or something

If you need completely black-box functions, you can have
```julia
struct BBConstraint{F, T}
    func::F
    rhs::T
end
is_valid(f::BBFunction, xvalues) = f.func(xvalues) <= rhs
```

The only thing I see is that this is basically re-implementing lots of things from MOI, so it could be relevant to start from what they have

Comment (Wikunia) [reply src/ConstraintSolver.jl]: The idea was that some constraints might share the same kind of constraint but are slightly different i.e have `==` vs. `<=` for the `LinearConstraint`. 
I prefer to have different functions for these cases instead of having `is_valid(lc, x)` and then check what the operator is in that function. If we have one `ConstraintType` for each constraint your idea would work but as some constraints share the same structure I found that wasteful.

Comment (matbesancon) [reply src/ConstraintSolver.jl]: you can then replace what I put above with a method `evaluate_at_value(func, xvalues)` or something similar. Every function should be able to evaluate itself at given values. Then what happens with the value (either <=, >= or something else), is up to the constraint type

Comment (Wikunia) [reply src/ConstraintSolver.jl]: Can you clarify what you mean by "evaluate"? 
Maybe the simple way of having a constraint.jl for each constraint works:
`LinearConstraint.jl`:
```
function prune!(com::CoM, lc::LinearConstraint)
      if lc.operator == :(==)
               return prune_eq_sum!(com, lc)
      end
      if lc.operator == :(<=)
               return prune_leq_sum!(com, lc)
     end
    ...
end
```
works and for the `BasicConstraint` we have an operator like `:alldifferent` or `:circuit` or something. If a new constraint gets added the developer needs to implement the normal prune functionality and then adds the `operator` to the constraint.jl
Then normal dispatch works + some `if` calls which should be okay performance wise, right?

Comment (matbesancon) [reply src/ConstraintSolver.jl]: I was thinking of the case when you have a function and a right-hand-side.`f(x) operator rhs`. You can make your function a type you dispatch on

**Comment by Wikunia** [comment]
Checked that there are no changes compared to master for killer sudoku regarding number of steps. The runtime might be slightly worse up but not significantly.

**Comment by Wikunia** [comment]
If there are no other things @matbesancon I would merge this.

**Comment by matbesancon** [comment]
Yup let's go with it :)

**Comment by matbesancon** [comment]
anyway tests are passing, future things to do can be in other PRs
